### ISSUE TITLE
Multitenant Quarkus application cannot access multiple tenant databases in one request

### ISSUE THREAD

**Issue by loban** [quarkusio/quarkus#33723]
I have implemented the following custom `TenantResolver` class.

```
@ApplicationScoped
@PersistenceUnitExtension
public class CustomTenantResolver implements TenantResolver {
    String tenantId;

    public void setTenantId(String tenantId) {
        this.tenantId = tenantId;
    }

    @Override
    public String getDefaultTenantId() {
        return "default";
    }

    @Override
    public String resolveTenantId() {
        return tenantId;
    }
}
```

And also the following custom `TenantConnectionResolver` class.

```
@ApplicationScoped
@PersistenceUnitExtension
public class CustomTenantConnectionResolver implements TenantConnectionResolver {
    public String resolveJdbcUrl(String tenantId) {
        return "jdbc:mysql://localhost:3306/db_" + tenantId
    }

    public AgroalDataSource resolveDataSource(String tenantId) {
        String jdbcUrl = resolveJdbcUrl(tenantId);

        AgroalDataSourceConfiguration ds = dataSource.getConfiguration();
        AgroalConnectionPoolConfiguration cp = ds.connectionPoolConfiguration();
        AgroalConnectionFactoryConfiguration cf = cp.connectionFactoryConfiguration();

        AgroalConnectionFactoryConfigurationSupplier cfSupplier
                = new AgroalConnectionFactoryConfigurationSupplier(cf).jdbcUrl(jdbcUrl);
        AgroalConnectionPoolConfigurationSupplier cpSupplier
                = new AgroalConnectionPoolConfigurationSupplier(cp).connectionFactoryConfiguration(cfSupplier);
        AgroalDataSourceConfigurationSupplier dsSupplier
                = new AgroalDataSourceConfigurationSupplier().connectionPoolConfiguration(cpSupplier);

        return AgroalDataSource.from(dsSupplier);
    }

    @Override
    public ConnectionProvider resolve(String tenantId) {
        AgroalDataSource tenantDataSource = resolveDataSource(tenantId);

        return new QuarkusConnectionProvider(tenantDataSource);
    }
}
```

And finally, the following service class.

```
public class CustomService {
    @Inject
    @PersistenceUnit
    CustomTenantResolver tenantResolver;

    @Inject
    @PersistenceUnit
    EntityManager entityManager;

    public something() {
        List<String> tenantIds = List.of("tenant1", "tenant2", "tenant3");

        for (String tenantId : tenantIds) {
            tenantResolver.setTenantId(tenantId);

            List<Entity> entities = entityManager.createQuery("FROM Entity", Entity.class).getResultList();

            System.out.println(entities);
        }
    }
}
```

However, when I run my Quarkus Application, I notice that in each iteration of the loop, **I get the results from the first tenant database**. In debugging I see that the TenantResolver and TenantConnectionResolver are correctly called with the right tenant id, but when the actual SQL call is made, **it seems to only use the first connection**. So all my results for each tenant is just the result of the first tenant repeated.

**Comment by loban** [comment]
I found this https://github.com/quarkusio/quarkus/issues/20533 and used a `ThreadLocal` like the example in that ticket, but the same result. Perhaps I'm doing something wrong, but I am unable to find any working example of accessing multiple tenants (with database multi-tenancy) in a single service method.

**Comment by yrodiere** [comment]
I suspect you're not using transactions? If so, the Hibernate ORM session is request-scoped, and indeed you'll only get one session per request, and that session will use whatever tenant ID you set before you first accessed Hibernate ORM.

Did you try running each unit of work in its own transaction, e.g. with [`QuarkusTransaction`](https://quarkus.io/guides/transaction#programmatic-approach)? Then each unit of work should get its own session and each should use its own tenant ID (provided you set it before starting the transaction).

**Comment by loban** [comment]
@yrodiere Thanks for your reply. Let me try your suggestion.

**Comment by loban** [comment]
@yrodiere So I **was** using transactions, but it **still** didn't work. In fact, I realized that even if I was just dealing with one tenant per request, it was returning wrong results (e.g. one tenant's data for another).

After much googling, I came across these tickets
- https://github.com/quarkusio/quarkus/issues/7019
- https://github.com/quarkusio/quarkus/issues/11729

And this code example
- https://github.com/HamburgChimps/quarkus-programmatic-datasource-configuration/blob/main/src/main/java/de/hamburgchimps/ProgrammaticDatasourceConfigurer.java

Where I realized that the problem was due to me dynamically creating datasources for multitenancy. When doing so, they don't automatically get registered in the TransactionManager. Once I did things exactly like the code example, everything worked!

I'll close this ticket.
